src/abstractdataframe/join.jl

```diff
@@ -85,27 +85,53 @@ function compose_joined_table(joiner::DataFrameJoiner, kind::Symbol,
     @assert nrow == length(all_orig_right_ixs) + loil
     ncleft = ncol(joiner.dfl)
     cols = Vector{Any}(ncleft + ncol(dfr_noon))
-    _similar = kind == :inner ? similar : similar_missing
+    # inner and left joins preserve non-missingness of the left frame
+    # it is also preserved if all right rows have left matches
+    _similar_left = kind == :inner || kind == :left || length(rightonly_ixs.join) == 0 ? similar : similar_missing
     for (i, col) in enumerate(columns(joiner.dfl))
-        cols[i] = _similar(col, nrow)
-        copy!(cols[i], view(col, all_orig_left_ixs))
+        on_col_ix = findfirst(joiner.left_on, names(joiner.dfl)[i])
+        if on_col_ix > 0 && kind == :right
+            # if right join, construct the on-column
+            # using the right frame to preserve missingness and cat.levels
+            rcol = joiner.dfr_on[on_col_ix]
+            cols[i] = similar(rcol, nrow)
+            copy!(cols[i], view(rcol, all_orig_right_ixs))
+            permute!(cols[i], right_perm)
```

Comment:
Isn't there an alternative solution which doesn't involve calling `permute!`? I'm concerned about the performance impact compared with the previous code which simply called `copy!`.

Reply:
This is the case only for the right joins. Previously it was handled later in the code (`if length(rightonly_ixs) > 0` block).
Actually, now it should be faster, because we compose the resulting column using the right frame directly, instead of initializing it using the left frame and fixing it later.